src/abi.md

```diff
@@ -1,21 +1,246 @@
+
+> [!NOTE]
+> With transitivity, this applies regardless of the mutability of either pointer type
+
+```rust
+unsafe fn foo(x: *mut u32){
+   unsafe{x.write(5);}
+}
+
+fn main(){
+    let f: unsafe fn(*mut ()) = unsafe{core::mem::transmute(foo as unsafe fn(_))}; // Type Erase the function
+    let mut val: u32 = 0;
+    let ptr = core::ptr::addr_of_mut!(val).cast::<()>(); // Get Opaque Userdata from somewhere
+    unsafe{f(ptr);}
+    assert_eq!(val, 5);
+}
+```
+
+r[abi.compatibility.reference-box]
+The types [`&T`], [`&mut T`], [`alloc::boxed::Box<T>`], and [`core::ptr::NonNull<T>`], are *abi compatible* with `*const T`
+
+> [!NOTE]
+> With transitivity, they are also *abi compatible* with each other, and with `*mut T`, as well as references/`Box` to different types that have the same *metadata type*.
+
+```rust
+fn foo(x: &mut i32){
+   *x = 5;
+}
+
+fn main(){
+    let f: unsafe fn(*mut ()) = unsafe{core::mem::transmute(foo as fn(_))}; // Type Erase the function
+    let mut val = 0;
+    let ptr = core::ptr::addr_of_mut!(val).cast::<()>(); // Get Opaque Userdata from somewhere
+    unsafe{f(ptr);}
+    assert_eq!(val, 5);
+}
+```
+
+r[abi.compatibility.core]
+The types [`core::mem::MaybeUninit<T>`], [`core::cell::UnsafeCell<T>`], and [`core::num::NonZero<T>`], are *abi compatible* with `T`.
+
+r[abi.compatibility.transparent]
+A [`struct`] declared with the `transparent` representation is *abi compatible* with its field that does not have size 0 and alignment 1, if such a field exists.
+
+r[abi.compatibility.zst]
+Two types, `T` and `U`, are *abi compatible* if both have size 0 and alignment 1.
+
+r[abi.compatibility.discriminant]
+If `T` is an a type listed in [layout.repr.rust.option.elision], and `U` is the type of the *elision candidate field*, then `T` is layout compatible with `U`.
+
+> [!NOTE]
+> `Option<U>`, `Result<U,Z>`, or `Result<Z,U>` are such types, when `U` are *elision candidate type*s, and `Z` is a 1-ZST type.
+>
+> Due to transitivity, two such types are *abi compatible* with each other if their *elision candidate field*s are *abi comaptible*
+
+r[abi.compatibility.fn-ptr]
+An [function pointer] type `T` is *abi compatible* with an [function pointer] type `U` if `T` and `U` have *abi compatible* tags.
+
+r[abi.compatibility.extern-tag]
+Two [abi tags][abi tag] are *abi compatible* if:
+* They are the same string, or
+* One tag is `"X"`, and the other is `"X-unwind"`
+
+> [!NOTE]
+> e.g. `extern "C"` and `extern "C-unwind"` are compatible with each other.
+
+r[abi.compatibility.signature]
+Two function signatures are compatible if:
+* The [abi tags][abi tag] of both signatures are *abi compatible*,
+* They have the same number of parameters, excluding C-varargs,
+* Each parameter of both signatures, in order, are *abi compatible*, and
+* Either both signatures have C-varargs, or neither signature does.
+
+> [!NOTE]
+> A signature is compatible with itself.
+
+r[abi.compatibility.simd-abi]
```

Comment:
Given that we consider this a bug and are very close to fixing the bug to at least emit warnings, I don't think we should list this here.

Reply:
This clause no longer exists in the library docs, so it should be removed here as well.

Reply:
Yep, was waiting for the stdlib to fix to land. Will fix this tomorrow.